diff --git a/envoy-launcher/main.go b/envoy-launcher/main.go
new file mode 100644
index 0000000..f19e364
--- /dev/null
+++ b/envoy-launcher/main.go
@@ -0,0 +1,74 @@
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"syscall"
+)
+
+const configTemplate = `{
+    "listeners": [{
+        "address": "tcp://0.0.0.0:%s",
+        "filters": [{
+            "type": "read",
+            "name": "tcp_proxy",
+            "config": {
+                "stat_prefix": "ingress_tcp",
+                "route_config": {
+                    "routes": [{
+                            "cluster": "service_local"
+                    }]
+                }
+            }
+        }],
+        "ssl_context": {
+          "cert_chain_file": "%s",
+          "private_key_file": "%s"
+        }
+    }],
+    "admin": {
+        "access_log_path": "/tmp/admin_access.log",
+        "address": "tcp://127.0.0.1:9901"
+    },
+    "cluster_manager": {
+        "clusters": [{
+            "name": "service_local",
+            "connect_timeout_ms": 250,
+            "type": "static",
+            "lb_type": "round_robin",
+            "hosts": [{
+                "url": "tcp://127.0.0.1:%s"
+            }]
+        }]
+    }
+}`
+
+const configFilePath = "/home/vcap/tmp/envoy-config.json"
+
+func main() {
+	certPath := os.Getenv("CF_INSTANCE_CERT")
+	keyPath := os.Getenv("CF_INSTANCE_KEY")
+	appPort := os.Getenv("PORT")
+
+	tlsListenPort := os.Getenv("TLS_PORT")
+	if tlsListenPort == "" {
+		tlsListenPort = "10000"
+	}
+
+	fullConfig := fmt.Sprintf(configTemplate, tlsListenPort, certPath, keyPath, appPort)
+	err := ioutil.WriteFile(configFilePath, []byte(fullConfig), os.ModePerm)
+	if err != nil {
+		log.Fatalf("writing config file: %s", err)
+	}
+
+	err = syscall.Exec("/tmp/lifecycle/envoy", []string{
+		"envoy",
+		"-c",
+		configFilePath,
+	}, os.Environ())
+	if err != nil {
+		log.Fatalf("exec: %s", err)
+	}
+}
